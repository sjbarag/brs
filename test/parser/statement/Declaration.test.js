const { Lexeme, Parser } = require("brs");
const BrsError = require("../../../lib/Error");

const { token, identifier, EOF } = require("../ParserTests");

describe("parser", () => {
    afterEach(() => BrsError.reset());

    describe("variable declarations", () => {
        it("allows newlines after assignments", () => {
            let parsed = Parser.parse([
                identifier("hasNewlines"),
                token(Lexeme.Equal),
                token(Lexeme.True),
                token(Lexeme.Newline),
                EOF
            ]);

            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
        });

        it("parses literal value assignments", () => {
            let parsed = Parser.parse([
                identifier("foo"),
                token(Lexeme.Equal),
                token(Lexeme.Integer, 5),
                EOF
            ]);

            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses evaluated value assignments", () => {
            let parsed = Parser.parse([
                identifier("bar"),
                token(Lexeme.Equal),
                token(Lexeme.Integer, 5),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 3),
                EOF
            ]);

            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses variable aliasing", () => {
            let parsed = Parser.parse([
                identifier("baz"),
                token(Lexeme.Equal),
                identifier("foo"),
                EOF
            ]);

            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });
});
