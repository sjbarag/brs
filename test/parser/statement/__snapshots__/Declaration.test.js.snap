// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser variable declarations parses evaluated value assignments 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "bar",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 5,
        },
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 3,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "Caret",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser variable declarations parses literal value assignments 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "foo",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 5,
      },
    },
  },
]
`;

exports[`parser variable declarations parses variable aliasing 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "baz",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "foo",
      },
      "type": "Variable",
    },
  },
]
`;
